tests/mock_crust/tunnel.rs

```diff
@@ -146,4 +169,104 @@ fn tunnel_clients() {
+        .is_none());
+}
+
+#[test]
+// The purpose of this test is to confirm that as in the logs of MAID-1951, once the tunnel got
+// established, a ConnectFailure of a tunnel_client to peer tunnel_client won't incur any action
+fn tunnel_client_connect_failure() {
+    let min_section_size = 5;
+    let network = Network::new(min_section_size, None);
+    network.block_connection(Endpoint(2), Endpoint(3));
+    network.block_connection(Endpoint(3), Endpoint(2));
+    let mut nodes = create_connected_nodes(&network, min_section_size);
+    let tunnel_node_index = unwrap!(locate_tunnel_node(&nodes, PeerId(2), PeerId(3)));
+
+    network.send_crust_event(Endpoint(2), crust::Event::ConnectFailure(PeerId(3)));
+    let _ = poll_all(&mut nodes, &mut []);
+    verify_invariant_for_all_nodes(&nodes);
+    assert_eq!(tunnel_node_index,
+               unwrap!(locate_tunnel_node(&nodes, PeerId(2), PeerId(3))));
+}
+
+fn verify_tunnel_switch(nodes: &mut Vec<TestNode>, node: usize, client_1: usize, client_2: usize) {
+    let mut event_count = 0;
+    while let Ok(event) = nodes[client_1].inner.try_next_ev() {
+        match event {
+            Event::NodeLost(name, _) => {
+                assert!(name == nodes[node].name() || name == nodes[client_2].name());
+                event_count += 1;
+            }
+            Event::NodeAdded(name, _) => {
+                assert!(name == nodes[client_2].name());
+                assert_eq!(event_count, 2);
+            }
+            _ => {
+                panic!("{:?} received unexpected event {:?}",
+                       nodes[client_1].name(),
+                       event)
+            }
+        }
+    }
+    event_count = 0;
+    while let Ok(event) = nodes[client_2].inner.try_next_ev() {
+        match event {
+            Event::NodeLost(name, _) => {
+                assert!(name == nodes[client_1].name());
+                event_count += 1;
+            }
+            Event::NodeAdded(name, _) => {
+                assert!(name == nodes[client_1].name());
+                assert_eq!(event_count, 1);
+            }
+            _ => {
+                panic!("{:?} received unexpected event {:?}",
+                       nodes[client_2].name(),
+                       event)
+            }
+        }
+    }
+}
+
+#[test]
+fn tunnel_node_disrupted() {
+    let min_section_size = 5;
+    let network = Network::new(min_section_size, None);
+    network.block_connection(Endpoint(2), Endpoint(3));
+    network.block_connection(Endpoint(3), Endpoint(2));
+    let mut nodes = create_connected_nodes(&network, min_section_size);
+    let tunnel_node_index = unwrap!(locate_tunnel_node(&nodes, PeerId(2), PeerId(3)));
+
+    network.send_crust_event(Endpoint(2),
+                             crust::Event::LostPeer(PeerId(tunnel_node_index)));
+    network.send_crust_event(Endpoint(tunnel_node_index),
+                             crust::Event::LostPeer(PeerId(2)));
```

Comment:
Yes, I think making that a method of the network itself would make sense. But it is not sufficient to just send the `LostPeer` events: it also needs to remove the connection entries from the `ServiceHandle`s, like here: https://github.com/maidsafe/routing/blob/master/src/mock_crust/support.rs#L452
So it probably needs to call `disconnect` on one of the service handles (which will automatically cause a `LostPeer` event in the other one), and then in addition raise the `LostPeer` on that side: https://github.com/maidsafe/routing/blob/master/src/mock_crust/support.rs#L476

Otherwise if the test e.g. tried to re-establish the connection, mock Crust would refuse to, because the connection already exists.